### ISSUE TITLE
[FR] Temporarily bring window to front as "Preview Hover Action"

### ISSUE THREAD

**Issue by Arnoud-B** [ejbills/DockDoor#248]
## Problem or Need
Need to see a preview of the original window in full resolution for readability, when hovering over different open application windows.
Currently, it shows a huge full-screen capture of the application window when the window is small. On a large screen, this doesn't make it much easier to read (everything blown up in size), and it's not looking great.

## Proposed Solution
Temporarily bring window to front while hovering over the preview in the dock, instead of showing a capture. Several other dock preview apps use this mechanism too.

## Potential Impact
Should save system resources as well (speedy, no need to update/use a capture).

## macOS Version
14.6.1

## App Version
1.1.5

- [x] I have reviewed the open issues to ensure this is not a duplicate feature request

**Comment by gkrawiec** [comment]
this is how hyperdock does them. It's pretty good.
I vote for this.

**Comment by ShlomoCode** [comment]
> Currently, it shows a huge full-screen capture of the application window when the window is small. On a large screen, this doesn't make it much easier to read (everything blown up in size), and it's not looking great.

Aren't you simply asking for an option to adjust the scaling ratio of the magnified preview displayed on hover over the normal preview?

**Comment by Arnoud-B** [comment]
@ShlomoCode That could perhaps also be a solution, as long as the preview is shown in exactly the same spot as the window (so when you click on the thumbnail, the window is shown in exactly the same spot as the preview in a seamless, fast transition).

**Comment by gkrawiec** [comment]
the way other apps did it was to show a mini screenshot preview of the windows while mouse hovered over the dock icons. If you moved the pointer over the previews and kept moving the mini previews were simply highlighted with a border based on where the pointer was. if you selected a preview, the full size window became the front window.
But, if you hovered over the preview, did not move and waited a second, it went into a mode were the full size live window came to the front but did not stay. As the pointer was moved from preview to preview the full size live window of the preview you were hovering on became the current front live window (live with video and everything). This continued until you either left the are of the dock and previews, in which case the the front window went back to whatever it was before you hovered over the previews, or if you clicked on a preview the now temporary live window became permanent.
Hopefully this made some sense.

**Comment by t1m0thyj** [comment]
+1 for this idea. DockDoor is an awesome app and by far the best free dock companion for MacOS that I've tried, The only thing it's missing for me is the ability to preview windows at their full size and actual screen position.

As a longtime Windows user and former user of the DockMate MacOS app, I've grown accustomed to this behavior. It's convenient especially when switching between multiple windows of the same program across multiple monitors to help distinguish them.

**Comment by ejbills** [comment]
I hear you guys! I just want to confirm the functionality: 

You want the hover preview to render on the SAME screen that the REAL window is on (rather than the same screen the preview is currently active on)? I have cooked up a work in progress, please let me know if this functionality is incorrect. 

_Note: this still relies on the image in the cache, so the resolution of your images must be high if you want the full resolution window previews as shown in the video._

https://github.com/user-attachments/assets/6116c111-aa08-4758-a6cd-e0c0f0300d17

**Comment by apotenza92** [comment]
Not sure about others but that looks like exactly what I want @ejbills you're an absolute wizard 🧙‍♂️

Take some time off to enjoy the holidays.

**Comment by gkrawiec** [comment]
No, not what I would expect. What I was hoping was not to show a larger preview. It was to actually open the app and bring it to tht front. Just as if I had selected it in the dock. The live app. The only difference with just opening it from the dock is that once I move away from the preview the app goes back to where it was (either minimized, behind other windows, etc). So if I just slide the pointer from preview to preview i can see each live app come in and go away.
I might be wrong, but I am pretty sure that is was hyperdock did.

> I hear you guys! I just want to confirm the functionality:
> 
> You want the hover preview to render on the SAME screen that the REAL window is on (rather than the same screen the preview is currently active on)? I have cooked up a work in progress, please let me know if this functionality is incorrect.
> 
> _Note: this still relies on the image in the cache, so the resolution of your images must be high if you want the full resolution window previews as shown in the video._
> window.hovering.action.preview.dev.mp4

**Comment by ejbills** [comment]
> No, not what I would expect. What I was hoping was not to show a larger preview. 
It isn't a larger preview, it is now sized exactly the same, and is placed in the exact position as the real window. 

> It was to actually open the app and bring it to tht front. Just as if I had selected it in the dock. The live app. The only difference with just opening it from the dock is that once I move away from the preview the app goes back to where it was (either minimized, behind other windows, etc). So if I just slide the pointer from preview to preview i can see each live app come in and go away.

I understand what you mean, however I'm not  confident the accessibility API allows you to tell an AXUIElement to go to a specific level within the window stack. You are able to use "kAXRaiseAction" to bring a window into focus, but there is no method of then telling that window to return to its original position. 

Along with this, what if said window is in another desktop space? It would switch your desktop space automatically, and then bring you back to the original desktop space after you are done hovering over that window?

At the moment, placing the window image that DockDoor has access to in the same position and dimensions as the *real* window exactly is the way to go for this feature. Manipulating the actual window element would add lots of additional complexity, for not much gained in terms of functionality, over just using the window image inside DockDoor. 

If you are concerned about the images not being updated, you can set the cache lifespan to 0, in which the preview shown will always be up to date (including minimized windows).

**Comment by gkrawiec** [comment]
maybe i recall the feature incorrectly, but I am pretty sure that is how it
worked.

Maybe sending it back to its location is not necessary, just bringing
bringing into focus
whatever is not in focus, and maximazing and then minimizing windows that
started minimized.
Of the ons you cant return to origins position, just leave in focus the
last one before you
leave the preview?
anyway, I am now used to using this app instead of hyperdock, so if not
possible,. not a problem.
i no longer miss the feature.


thanks for spending time on this!
